lib/properties.ts

```diff
@@ -28,24 +28,31 @@ import path from 'path';
 import _ from 'underscore';
 
 import {logger} from './logger';
+import {PropertyGetter, PropertyValue, Widen} from './properties.interfaces';
 import {toProperty} from './utils';
 
-let properties = {};
+let properties: Record<string, Record<string, PropertyValue>> = {};
 
-let hierarchy = [];
+let hierarchy: string[] = [];
 
 let propDebug = false;
 
-function findProps(base, elem) {
-    const name = base + '.' + elem;
-    return properties[name];
+function findProps(base: string, elem: string): Record<string, PropertyValue> {
+    return properties[`${base}.${elem}`];
 }
 
 function debug(string) {
     if (propDebug) logger.info(`prop: ${string}`);
 }
 
-export function get(base, property, defaultValue) {
+export function get(base: string, property: string, defaultValue: undefined): PropertyValue;
+export function get<T extends PropertyValue>(
+    base: string,
+    property: string,
+    defaultValue: Widen<T>,
+): typeof defaultValue;
+export function get<T extends PropertyValue>(base: string, property: string, defaultValue?: unknown): T;
+export function get(base: string, property: string, defaultValue?: unknown): unknown {
```

Comment:
The final 👨🏻‍🍳 💋 would be to check the return value really is a `T` ... but... I don't know if that's possible in TS.

Reply:
I was thinking the same, it'd be really nice to enforce this. I think it is possible if we require a default argument always be passed, but I'm not sure how we'd handle cases were we expect, e.g., `string | undefined` from the Props.

Alternatively some day we could make the props interface look like this
```ts
props.getString("key");
props.getNumber("key");
props.getBool("key");
props.getOptionalString("key");
props.getOptionalNumber("key");
props.getOptionalBool("key");
```